### PR TITLE
backupccl: elide data from offline tables during RESTORE

### LINKED ISSUE #86054: backupccl: incrementally backup in progress imports on existing tables and elide importing data in RESTORE
Currently, a table with an in-progress cannot get backed up, let-alone restored to its pre-import state.  Backing up an in-progress import also has the benefit of distributing the work of backing up the import over a series of incremental backups, as oppose to what currently occurs: the first incremental backup to begin after the import finishes has to back up everything.

To address this, main challenge involves rolling back imported data on the restored cluster. To understand why this is a challenge, consider how rollbacks occur today:

If an IMPORT writing data into an existing, non-empty cluster fails or is cancelled mid-IMPORT, to roll it back, any rows it had written are found and deleted by scanning the table for rows with a timestamp greater than the time at which the IMPORT started. This works since the table is offline to other writes while it is importing, but relies on the fact that the timestamps on rows do not change -- which may not be true if the table were backed up and then restored, after which all keys, both existing and imported, would have times based on when it was restored. 

The second paragraph in #76722 outlined one strategy which involved writing additional metadata to each imported key, and indeed several PRs began implementing this approach (#85338, #85692 #85138). However, we realized that binding the Import Start Time in the backed up table descriptor is sufficient. Specifically, when the `restore_data_processor` rewrites backed up keys to the restore cluster, it can use the ImportStartTime in the restored table descriptor to filter out keys in the backed up, in-progress import, _before_ AddSSTable rewrites the timestamps of all the keys.

Note: the more complicated approach outlined in #76722 would have been necessary if RESTOREs of _whole_ tenants implemented MVCC AddSSTable-- i.e. rewrote timestamps in RESTORE--  because during the restore, the host tenant cannot access tenant table descriptors and thus filter keys in the restore processor. And indeed, we thought it was necessary to make _whole_ tenant RESTOREs MVCC compatible. But now, we no longer think that whole tenant operations (like tenant streaming) need to be MVCC, since it's relatively easy to ensure that all downstream operations understand that whole tenant operations are non-MVCC. So given that whole tenant restores will continue to preserve timestamps from the backup, the restored tenant can rollback their import using the normal process described in the second paragraph.

Jira issue: CRDB-18546

### PR DESCRIPTION
**Pull request by msbutler**
A previous PR (https://github.com/cockroachdb/cockroach/pull/83915) enabled BACKUP to target offline descriptors
(including IMPORTing and RESTOREing tables), but didn't fully handle their
rollback on RESTORE. This PR ensures RESTORE properly handles these offline
tables:
- An offline table created by RESTORE or IMPORT PGDUMP is fully discarded. The
  table does not exist in the restoring cluster.
- An offline table created by an IMPORT INTO has all importing data elided in
  the restore processor and is restored online to its pre import state.

If the RESTORE occurs AOST after the table returns online, the table is
RESTOREd to its post offline state.

Currently, all tables that return online are fully re-backed up. For many
situations above, this is unecessary. Future work should ensure that these
tables are not re-backed up when we know all operations within the
incremental backup interval on the table were MVCC.

This patch only applies to BACKUP and RESTORE flavors that implicitly target an
offline importing/restoring table. A future PR should allow a user to
explicitly target an existing offline table undergoing an IMPORT.

Fixes #86054

Release justification: fixes bug where offline tables were not properly handled
in restore.

### CODE REVIEW

**Comment by msbutler** [comment]
RFAL! all test failures relate to first commit which relates to #85979, and not this PR.

File: pkg/ccl/backupccl/key_rewriter.go [hunk 1]
```diff
@@ -258,23 +258,33 @@ func (kr *KeyRewriter) RewriteKey(key []byte) ([]byte, bool, error) {
...
 	_, tableID, _ := keys.SystemSQLCodec.DecodeTablePrefix(key)
+
+	desc := kr.descs[descpb.ID(tableID)]
+	if desc == nil {
+		return nil, false, errors.Errorf("missing descriptor for table %d", tableID)
+	}
+
+	// Elide Importing Keys
+	if importTime := desc.GetInProgressImportStartTime(); wallTime > 0 && importTime > 0 && wallTime >= importTime {
+		return nil, false, ErrImportingKeyError
```
Comment (dt) [inline pkg/ccl/backupccl/key_rewriter.go:281]: do you need a non-nil err? isn't false enough?

**Comment by msbutler** [comment]
RFAL. I limited the scope of this PR to only deal with handling offline tables in restore. Currently on the backup side, all tables that come back online will get fully re-backed up. In a future PR, I will skip this re-backup step one tables that did not undergo a non-mvcc op during the inc backup interval.

File: pkg/ccl/backupccl/restore_job.go [hunk 2]
```diff
@@ -657,6 +657,25 @@ func shouldPreRestore(table *tabledesc.Mutable) bool {
 	return ok
 }
 
+// eligibleInProgressImport returns true if the descriptor represents a table with an in progress
+// import that started in a cluster finalized to version 22.2.
```
Comment (msbutler) [inline pkg/ccl/backupccl/restore_job.go:661]: nit: should say: "import that started ina cluster finalized to version 22.2 and is getting restored on finalized 22.2 cluster"

Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go:661]: the restoring cluster version check is superfluous as the descriptor's start time will never be above 0 in a mixed version restore. Perhaps I should instead conduct an assertion failure after checking that the import start time is greater than 0.

File: pkg/ccl/backupccl/restore_job.go [hunk 3]
```diff
@@ -660,6 +660,27 @@ func shouldPreRestore(table *tabledesc.Mutable) bool {
 	return ok
 }
 
+// eligibleInProgressImport returns true if the descriptor represents a table with an in progress
```
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_job.go:663]: nit: can we maybe clarify that this method is used on "backed up" descriptors and not descriptors in the cluster. To that effect what do you think about renaming this to `backedUpDescriptorsWithInProgressImportInto`, a bit wordy but maybe its okay.

File: pkg/ccl/backupccl/restore_job.go [hunk 4]
```diff
@@ -660,6 +660,27 @@ func shouldPreRestore(table *tabledesc.Mutable) bool {
...
+) (bool, error) {
+	table, ok := desc.(catalog.TableDescriptor)
+	if !ok {
+		return false, nil
+	}
+
+	if table.GetInProgressImportStartTime() == 0 {
+		return false, nil
+	}
+	if !p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.Start22_2) {
```
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_job.go:676]: As we discussed on our call, I think this version check is a little disconnected from what it is guarding against. If we don't want descriptors to have a GetInProgressImportStartTime set in a mixed version we should guard it where we set it (I think you already do), instead of where we read it. I would delete this but if you feel better with it here its okay too.

File: pkg/ccl/backupccl/restore_job.go [hunk 5]
```diff
@@ -95,7 +95,7 @@ var restoreStatsInsertBatchSize = 10
 func rewriteBackupSpanKey(
 	codec keys.SQLCodec, kr *KeyRewriter, key roachpb.Key,
 ) (roachpb.Key, error) {
-	newKey, rewritten, err := kr.RewriteKey(append([]byte(nil), key...))
+	newKey, rewritten, err := kr.RewriteKey(append([]byte(nil), key...), 0)
```
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_job.go:98]: nit: /* wallTime */

File: pkg/ccl/backupccl/restore_data_processor.go [hunk 6]
```diff
@@ -464,7 +464,12 @@ func (rd *restoreDataProcessor) processRestoreSpanEntry(
 		valueScratch = append(valueScratch[:0], iter.UnsafeValue()...)
 		value := roachpb.Value{RawBytes: valueScratch}
 
-		key.Key, ok, err = kr.RewriteKey(key.Key)
+		key.Key, ok, err = kr.RewriteKey(key.Key, key.Timestamp.WallTime)
+
+		if errors.Is(err, ErrImportingKeyError) {
+			// Elide importing keys
```
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_data_processor.go:470]: Lets add more to this comment, the logic is subtle so itll be nice to call out what and why we are eliding keys with this error.

File: pkg/ccl/backupccl/key_rewriter.go [hunk 7]
```diff
@@ -258,23 +258,33 @@ func (kr *KeyRewriter) RewriteKey(key []byte) ([]byte, bool, error) {
...
-// with the system codec. It is the responsibility of the caller to either
-// remap, or re-prepend any required tenant prefix.
-func (kr *KeyRewriter) rewriteTableKey(key []byte) ([]byte, bool, error) {
+// ErrImportingKeyError indicates the current key is apart of an in-progress import
+var ErrImportingKeyError = errors.New("Importing Key")
+
+// checkAndRewriteTableKey rewrites the table IDs in the key. It assumes that
+// any tenant ID has been stripped from the key so it operates with the system
+// codec. It is the responsibility of the caller to either remap, or re-prepend
+// any required tenant prefix.
```
Comment (adityamaru) [inline pkg/ccl/backupccl/key_rewriter.go:267]: Can we add a few sentences about the new parameter we have added, and what its used for.

File: pkg/ccl/backupccl/key_rewriter.go [hunk 8]
```diff
@@ -258,23 +258,33 @@ func (kr *KeyRewriter) RewriteKey(key []byte) ([]byte, bool, error) {
...
 	// they will be caught later on if tableID isn't in descs or kr doesn't
 	// perform a rewrite.
 	_, tableID, _ := keys.SystemSQLCodec.DecodeTablePrefix(key)
+
+	desc := kr.descs[descpb.ID(tableID)]
+	if desc == nil {
+		return nil, false, errors.Errorf("missing descriptor for table %d", tableID)
+	}
+
+	// Elide Importing Keys
```
Comment (adityamaru) [inline pkg/ccl/backupccl/key_rewriter.go:279]: Lets add more to this comment too, in particular the timestamp comparison and why its what we want in a restore.

File: pkg/ccl/backupccl/key_rewriter.go [hunk 9]
```diff
@@ -258,23 +258,33 @@ func (kr *KeyRewriter) RewriteKey(key []byte) ([]byte, bool, error) {
 	return rekeyed, ok, err
 }
 
-// rewriteTableKey rewrites the table IDs in the key.
-// It assumes that any tenant ID has been stripped from the key so it operates
-// with the system codec. It is the responsibility of the caller to either
-// remap, or re-prepend any required tenant prefix.
-func (kr *KeyRewriter) rewriteTableKey(key []byte) ([]byte, bool, error) {
+// ErrImportingKeyError indicates the current key is apart of an in-progress import
+var ErrImportingKeyError = errors.New("Importing Key")
```
Comment (adityamaru) [inline pkg/ccl/backupccl/key_rewriter.go:262]: nit: I think all errors should be lower case unless its a proper noun, maybe "skipping rewrite of an importing key"?

File: pkg/ccl/backupccl/key_rewriter.go (hunk 7 above)
Comment (adityamaru) [reply pkg/ccl/backupccl/key_rewriter.go:267]: Also maybe mention that a zero value skips the checks associated with the wallTime.

**Review by adityamaru — COMMENTED** [review]
Flushing out one round of comments, I need to look more carefully at the tests but can we maybe write a unit test in `key_rewriter_test.go` that isolates the new KR logic? The restore/import tests are super valuable but a unit test will give us more focussed control on testing the timestamp comparisons.

File: pkg/ccl/backupccl/restore_job.go [hunk 10]
```diff
@@ -708,6 +729,17 @@ func createImportingDescriptors(
 	preRestoreTables := make([]catalog.TableDescriptor, 0)
 
 	for _, desc := range sqlDescs {
+
+		// TODO (msbutler) remove the schema_change condition once the online schema changer
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go:736]: Done.

File: pkg/ccl/backupccl/restore_job.go [hunk 11]
```diff
@@ -708,6 +729,17 @@ func createImportingDescriptors(
...
+		// doesn't rely on OFFLINE state
+		if desc.Offline() && desc.GetDeclarativeSchemaChangerState() == nil {
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go:738]: Done.

File: pkg/ccl/backupccl/restore_job.go (hunk 3 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go]: Done.

File: pkg/ccl/backupccl/restore_job.go (hunk 4 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go]: Done. Deleted.

**Review by msbutler — COMMENTED** [review]
thanks for first pass! I think i got all your comments and added a keyrewriter test as well.

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+#######################
+
+# In an unfinalized cluster, back up some in-progress imports, and ensure that once the tables come
+# back online we fully back them again.
+#
+# Note that during IMPORT planning on an unfinalized cluster, the
+# ImportStartTime is not bound to the table's descriptor, therefore during
+# RESTORE AOST in-progress IMPORT, these tables should get thrown out.
+
+# TODO (msbutler): Currently, we still incrementally back up data from these
```
Comment (adityamaru) [inline pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:228]: what jobs are we referring to here?

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+exec-sql
+BACKUP DATABASE d INTO 'nodelocal://0/database_upgrade/' with revision_history;
+----
+
+
+save-cluster-ts tag=m0
+----
+
+
+# Conduct another set of incremental backups to ensure that _offline_ tables do _not_ fully backup
```
Comment (adityamaru) [inline pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:291]: can you clarify what you mean by "fully backup" here?

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports [hunk 14]
```diff
@@ -0,0 +1,469 @@
+# This test ensures that database and cluster backups properly
```
Comment (adityamaru) [inline pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:1]: meta note: we're starting 6 servers in this test which could make it quite a slow test, I don't know if theres anything we can do about it given the lack of a "wipe everything" tool for cluster restore but we should run this under stress a few times.

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+WHERE
+  object_name = 'foo' or object_name = 'foofoo';
+----
+d foo table 1 full
+d foofoo table 2 full
+d foo table 2 incremental
+d foofoo table 3 incremental
+
+
+# A restore on mixed version cluster should always fail on a backup that contains import times
```
Comment (adityamaru) [inline pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:421]: The backup to database_upgrade won't contain descriptors with import time set right even though its post finalization, because the import was run in a pre-finalized state?

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+
+
+exec-sql expect-error-regex='backup from version 22.1-2 is newer than current version'
+RESTORE DATABASE FROM LATEST IN 'nodelocal://0/database_upgrade/';
+----
+regex matches error
+
+
+# Now actually Restore the backups taken from a mixed version chain
+new-server name=s6 share-io-dir=s1 allow-implicit-access
```
Comment (adityamaru) [inline pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:433]: Can you just upgrade the previous server and reuse it, we can save a server startup I think.

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+----
+
+
+# Ensure the RESTOREs omit the tables with in progress imports (foo and foofoo)
+# as their descriptor will not have the import start time.
+restore aost=m0
+RESTORE FROM LATEST IN 'nodelocal://0/cluster/' AS OF SYSTEM TIME m0;
```
Comment (adityamaru) [inline pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:440]: can we check RESTORE FROM DATABASE AOST m0 as well after dropping d?

**Review by adityamaru — CHANGES_REQUESTED** [review]
Nice test coverage, and thanks for the comments!

File: pkg/ccl/backupccl/key_rewriter.go [hunk 18]
```diff
@@ -276,7 +288,12 @@ func (kr *KeyRewriter) checkAndRewriteTableKey(key []byte, wallTime int64) ([]by
 		return nil, false, errors.Errorf("missing descriptor for table %d", tableID)
 	}
 
-	// Elide Importing Keys
+	// If the user passes a non-zero walltime for the key, and the key's table is
+	// undergoing an IMPORT (indicated by a non zero
+	// GetInProgressImportStartTime), then this function returns an error if this
+	// key is a part of the import -- i.e. the key's walltime is greater than the
+	// import start time. It is up to the caller to handle this error properly.
+	fmt.Printf("walltime: %v; importTime %v \n", wallTime, desc.GetInProgressImportStartTime())
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/key_rewriter.go:296]: Ah nice catch. Removed.

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go [hunk 19]
```diff
@@ -619,7 +619,7 @@ func (sip *streamIngestionProcessor) bufferSST(sst *roachpb.RangeFeedSSTable) er
 }
 
 func (sip *streamIngestionProcessor) rekey(key roachpb.Key) ([]byte, error) {
-	rekey, ok, err := sip.rekeyer.RewriteKey(key)
+	rekey, ok, err := sip.rekeyer.RewriteKey(key, 0 /*wallTime*/)
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:622]: Currently, passing a 0 walltime signals to RewriteKey to _never_ filter on walltime. A SIP may encounter a descriptor with an in progress import, and I don't want the SIP do key filtering. Passing a 0 walltime signals this. We could easily rename the var `walltimeFilter` to make this slightly more readable.

Alternatively, the caller could always pass a non-zero walltime and _also_ pass a `filterbyWalltime` bool to `makeKeyRewriter` that binds to the `KeyRewriter` struct which signals whether walltimes should be used for filtering in `RewriteKey`.

Again, I prefer what I've initially drafted, as it requires managing fewer internal vars and has no perf impact on SIP. It's up to the caller to know whether they want to filter on import start time via this low level function thats optimized for performance.

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-restores [hunk 20]
```diff
@@ -0,0 +1,132 @@
+# This test ensures that database and cluster backups properly
+# backup and restore an in progress RESTORE
+#
+new-server name=s1
+----
+
+exec-sql
+CREATE DATABASE b;
+USE b;
+CREATE TABLE baz (i INT PRIMARY KEY, s STRING);
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/testdata/backup-restore/in-progress-restores:10]: Done. Let me know if this aligns with what you had in mind!

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (adityamaru) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:622]: > I don't want the SIP do key filtering

Have we thought this through though, as to why we don't want this behaviour to be uniform with the SIP? I am not very  familiar with the implications of this so @stevendanna can probably help make a decision here.

> It's up to the caller to know whether they want to filter on import start time

I agree with this sentiment but as you know I don't think this code is a reflection of that. The filtering should not be in this method but given our offline discussions, I think a happy compromise would be to unconditionally pass in the ts of the key that we are passing in if the semantics are well understood for the SIP.

Renaming the variable to something other than `timestamp` makes it even more likely that we pass in a timestamp that is not the keys timestamp, which we should never iiuc?

**Review by msbutler — COMMENTED** [review]
you're welcome for all the new test coverage :D

**Comment by msbutler** [comment]
@adityamaru RFAL!

**Review by adityamaru — APPROVED** [review]
LGTM with nits. Thanks for working with me through this, I think the test coverage is extensive so we should get this in and let it bake. Can you ping SQL Schema once this diff is borsing so that they are well informed when they're fixing their bug?

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-restores [hunk 21]
```diff
@@ -125,8 +147,295 @@ b2
...
+----
+
+
+restore expect-pausepoint tag=b
+RESTORE TABLE b.me.baz FROM LATEST IN 'nodelocal://0/cluster/' WITH into_db='d';
+----
+job paused at pausepoint
+
+
+# ensure user cannot override offline schema created by table restore into data database d
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/testdata/backup-restore/in-progress-restores:185]: Ah good catch-- though I'm not sure d in the backed up cluster from L139 actually has the schema. Anyway, I added the following to clarify things, and this test still holds.
```
USE d;
DROP SCHEMA IF EXISTS me CASCADE;
```


Note that during  the beginning of restore execution, the schema is published to an offline state. I'm not sure how to implement the UX you want. How is the user supposed to override an already published schema? Seems like an issue for another day.

File: pkg/ccl/backupccl/key_rewriter_test.go [hunk 22]
```diff
@@ -147,6 +147,40 @@ func TestKeyRewriter(t *testing.T) {
...
+		newKr, err := MakeKeyRewriterFromRekeys(keys.SystemSQLCodec, []execinfrapb.TableRekey{
+			{OldID: uint32(oldID), NewDesc: mustMarshalDesc(t, desc.TableDesc())},
+		}, nil /* tenantRekeys */, false /* restoreTenantFromStream */)
+		require.NoError(t, err)
+
+		key := rowenc.MakeIndexKeyPrefix(keys.SystemSQLCodec,
+			systemschema.NamespaceTable.GetID(), desc.GetPrimaryIndexID())
+
+		// If the passed in walltime is at or above the ImportStartWalltime, an error should return
+		_, _, err = newKr.RewriteKey(key, 2)
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/key_rewriter_test.go:164]: Done.

File: pkg/ccl/backupccl/restore_job.go [hunk 23]
```diff
@@ -718,6 +718,10 @@ func createImportingDescriptors(
 
 	oldTableIDs := make([]descpb.ID, 0)
 
+	// offlineSchemas are used to ensure that restoring databases descriptors do
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go]: Done.

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-restores [hunk 24]
```diff
@@ -53,43 +68,50 @@ SET CLUSTER SETTING jobs.debug.pausepoints = '';
...
+b2 public schema <nil> incremental
+b2 me schema <nil> incremental
+b2 greeting type <nil> incremental
+b2 _greeting type <nil> incremental
+b2 baz table 3 incremental
+b2 public schema <nil> incremental
+b2 me schema <nil> incremental
+b2 greeting type <nil> incremental
+b2 _greeting type <nil> incremental
 b2 baz table 3 incremental
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/testdata/backup-restore/in-progress-restores:114]: Answered offline.

**Comment by msbutler** [comment]
TFTRS!!

bors r=adityamaru
